urldee901fae6_DocX_personal_timetabling_tgz-pJ8-net_personaltt_main_BenchmarkJ8/src/net/personaltt/solver/heuristics/RouletteOccurrenceSelection.java

```diff
@@ -0,0 +1,112 @@
+/*
+ * To change this template, choose Tools | Templates
+ * and open the template in the editor.
+ */
+package net.personaltt.solver.heuristics;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Random;
+import net.personaltt.model.Occurrence;
+import net.personaltt.model.Schedule;
+import net.personaltt.solver.core.OccurrenceSelection;
+import net.personaltt.solver.core.SolverState;
+import net.personaltt.utils.intervalmultimap.IntervalMultimap;
+import net.personaltt.utils.RandomUtils;
+import net.personaltt.utils.ValuedInterval;
+import net.sf.cpsolver.ifs.util.DataProperties;
+
+/**
+ * Roulette occurrence selection. Selects occurrence with larger cost with more probability
+ * cost is sum of lengths shared with occurrence for each of other occurrences allocations
+ * @author docx
+ */
+public class RouletteOccurrenceSelection implements OccurrenceSelection {
+
+    Random random = new Random();   
+    
+    double optimizationWhenConflictProb;
+
+    public RouletteOccurrenceSelection(DataProperties properties) {
+        // EXPERIMENT with added probability of also counting preference cost when
+        // conflict solving phase. According to experiments, zero is best so far.
+        optimizationWhenConflictProb = 
+                properties.getPropertyDouble("rouletteOccurrenceSelection.optimizationWhenConflictProb", 0);
+    }
+    
+     /**
+     * Gets occurrence to solve using heuristic from conflicting occurrences with
+     * cost
+     * @param conflictingOccurrencesWithCost
+     * @return ConflictSumAllocationCost
+     */    
+    @Override
+    public Occurrence select(SolverState solution) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 12. By comparison, 98% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code block on lines 49-107 into a separate method.